keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,539 @@
+- NodeGetVolumeStats RPC
+  - For any PV that is mounted, the external agent calls `NodeGetVolumeStats` to see if volume is still mounted;
+  - Calls `NodeGetVolumeStats` to check if volume is usable, e.g., filesystem corruption, bad blocks, etc.
+
+Two new controller capabilities are added. If a CSI driver supports `GET_VOLUME` capability, it MUST support calling `GetVolume` to provide general volume information in `GetVolumeResponse`. If a driver supports `GET_VOLUME_HEALTH`, it MUST provide additional health information in `GetVolumeResponse`.
+
+A new node capability is added. If CSI driver supports the `GET_VOLUME_STATS_HEALTH` capability, it MUST provide health information in `NodeGetVolumeStats`.
+
+Detailed changes needed in the CSI Spec are described in the following.
+
+#### Add GetVolume RPC
+
+Add a new `GetVolume` RPC. `GetVolumeResponse` should contain current information of a volume if it exists. If the volume does not exist any more, `GetVolume` should return gRPC error code `NOT_FOUND`.
+
+Add `VolumeHealth` message in `VolumeStatus` field in `GetVolumeResponse`. This can indicate additional health information for the volume.
+
+Whether the volume is still attached to a node is already handled by `LIST_VOLUMES_PUBLISHED_NODES` controller capability. This capability will be used by Kubernetes to reconcile - re-attach the volume if the volume is not attached any more.
+
+```
+message GetVolumeRequest {
+  // Identity information for a specific volume. This field is
+  // REQUIRED. GetVolume will return with current volume information.
+  string volume_id = 1;
+}
+```
+
+```
+message GetVolumeResponse {
+  message VolumeStatus{
+    // health shows error conditions reported by the SP.
+    // This field MUST be specified if the
+    // GET_VOLUME_HEALTH controller capability is supported.
+    VolumeHealth health = 1;
+  }
+
+  // This field is REQUIRED
+  Volume volume = 1;
+
+  // This field is OPTIONAL. This field MUST be specified if the
+  // GET_VOLUME_HEALTH controller capability is supported.
+  VolumeStatus status = 2;
+}
+```
+
+```
+message VolumeHealth {
+  // The error code describing the health condition of the volume.
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error_code. This field is OPTIONAL.
+  string message = 2;
+}
+```
+
+The following common error codes are proposed for volume health:
+* VolumeNotFound
+* VolumeUnmounted
+* OutOfCapacity
+* RWIOError
+
+There could be many other error conditions, i.e. DiskFailure, DiskRemoved, ConfigError, NetworkError, DiskDegrading, FilesystemCorruption, etc. Because different storage systems may report health status differently, we leave it to the CSI drivers to determine what other error codes to report.
```

Comment:
Are we planning to introduce these other error codes in a follow up PR?

Reply:
That depends on whether sig-storage wants to add them in the future.  So we can start with a minimal set that we can agree upon, and add more when needed.

Reply:
These sound like useful and common error conditions to surface too, even if we don't use it in Kubernetes. Can we include all these other error codes as well to help avoid fragmentation?

Reply:
Sure